src/fast_layers.jl

```diff
@@ -24,85 +24,212 @@ initial_params(c::FastChain) = vcat(initial_params.(c.layers)...)
+  if typeof(f.cache) <: Nothing
+    if !isgpu(p)
+      W = @view(p[reshape(1:(f.out*f.in),f.out,f.in)])
     else
-      zbar = ȳ .* ForwardDiff.derivative.(f.σ,r)
+      W = reshape(@view(p[1:(f.out*f.in)]),f.out,f.in)
     end
-    Wbar = zbar * x'
-    bbar = zbar
-    xbar = W' * zbar
-    pbar = if f.bias == true
-        tmp = typeof(bbar) <: AbstractVector ?
-                         vec(vcat(vec(Wbar),bbar)) :
-                         vec(vcat(vec(Wbar),sum(bbar,dims=2)))
-        ifgpufree(bbar)
-        tmp
+    if f.bias == true
+      b = p[(f.out*f.in + 1):end]
+      r = W*x .+ b
+      ifgpufree(b)
     else
-        vec(Wbar)
+      r = W*x
+    end
+    y = f.σ.(r)
+  else
+    if !isgpu(p)
+      f.cache.W .= @view(p[reshape(1:(f.out*f.in),f.out,f.in)])
+    else
+      f.cache.W .= reshape(@view(p[1:(f.out*f.in)]),f.out,f.in)
+    end
+    f.cache.cols[1] = size(x)[2]
+    mul!(@view(f.cache.r[:,1:f.cache.cols[1]]), f.cache.W, x)
+    if f.bias == true
+      @view(f.cache.r[:,1:f.cache.cols[1]]) .+= @view(p[(f.out*f.in + 1):end])
     end
-    ifgpufree(Wbar)
-    ifgpufree(r)
-    nothing,xbar,pbar
+    @view(f.cache.y[:,1:f.cache.cols[1]]) .= f.σ.(@view(f.cache.r[:,1:f.cache.cols[1]]))
+  end
+  function FastDense_adjoint(ȳ)
+    if typeof(f.cache) <: Nothing
+      if typeof(f.σ) <: typeof(tanh)
+        zbar = ȳ .* (1 .- y.^2)
+      elseif typeof(f.σ) <: typeof(identity)
+        zbar = ȳ
+      else
+        zbar = ȳ .* ForwardDiff.derivative.(f.σ,r)
+      end
+      Wbar = zbar * x'
+      bbar = zbar
+      xbar = W' * zbar
+      pbar = if f.bias == true
+          tmp = typeof(bbar) <: AbstractVector ?
+                           vec(vcat(vec(Wbar),bbar)) :
+                           vec(vcat(vec(Wbar),sum(bbar,dims=2)))
+          ifgpufree(bbar)
+          tmp
+      else
+          vec(Wbar)
+      end
+      ifgpufree(Wbar)
+      ifgpufree(r)
+      nothing,xbar,pbar
+    else
+      if typeof(f.σ) <: typeof(tanh)
+        @view(f.cache.zbar[:,1:f.cache.cols[1]]) .= ȳ .* (1 .- @view(f.cache.y[:,1:f.cache.cols[1]]).^2)
+      elseif typeof(f.σ) <: typeof(identity)
+        @view(f.cache.zbar[:,1:f.cache.cols[1]]) .= ȳ
+      else
+        @view(f.cache.zbar[:,1:f.cache.cols[1]]) .= ȳ .* ForwardDiff.derivative.(f.σ, @view(f.cache.r[:,1:f.cache.cols[1]]))
+      end
+      mul!(f.cache.Wbar, @view(f.cache.zbar[:,1:f.cache.cols[1]]), x')
+      mul!(@view(f.cache.xbar[:,1:f.cache.cols[1]]), f.cache.W', @view(f.cache.zbar[:,1:f.cache.cols[1]]))
+      f.cache.pbar .= if f.bias == true
+        vec(vcat(vec(f.cache.Wbar),sum(@view(f.cache.zbar[:,1:f.cache.cols[1]]),dims=2)))# bbar = zbar
+      else
+        vec(f.cache.Wbar)
+      end
+      nothing,@view(f.cache.xbar[:,1:f.cache.cols[1]]),f.cache.pbar
+    end
+  end
+  if typeof(f.cache) <: Nothing
+    y,FastDense_adjoint
+  else
+    @view(f.cache.y[:,1:f.cache.cols[1]]),FastDense_adjoint

```

Comment:
have an `out` array that you write into and then return, instead of returning a view which will allocate the mutable struct of the view itself. That's probably the 176.

Reply:
can we do anything for this view without the confirmation that numcols will equal cols?

Reply:
can we do anything for this view without the confirmation that numcols will equal cols?

Reply:
we can return the whole y if numcols equals cols and if cols is lesser, the view